### PR TITLE
Extend code length to 64 bits and allow single pulses

### PR DESCRIPTION
**Pull request by lulol**
- Extend code length to 64 bits (unsigned long long)
- Allow single pulses (pulses.high or pulses.low == 0)
- Add raw protocol for TR-502MSV/RC701 LIDL switches. i.e.:
  
  ch0.on: 001 011 001 001 001 001 011 011 011 011 011 011 011 011 011 011 001 011 011 011 001 0
  ch0.off: 001 011 001 001 001 001 011 011 011 011 011 011 011 011 011 011 011 011 011 011 011 0

### CODE REVIEW

**Comment by fingolfin** [comment]
Single pulses don't really make sense. Perhaps what you is to swap high and low, as in PR #49 ?

**Comment by fingolfin** [comment]
To elaborate: Single pulses don't make sense because a "pulse" only occurs if a signal level changes.

So it looks as if you are abusing the system to bypass the encoding of ones and zeros in terms of pulse lengths. That also explains why you have such long raw packets: Every 001  corresponds to a 2-1 pulse, every 011 to a 1-2 pulse.

Did you try using a variation of protocol 2?

**Comment by SebastianJuch** [comment]
Hi, is it possible to extend the Code to 112 bits + 2 for Sync ?

**Comment by lulol** [comment]
@SebastianJuch Not as trivially as was to just use `unsigned long long` to extend it to 64 bit.
As fas as I'm aware, GCC's __int128 is only supported for 64bit architectures if I'm not mistaken.

Right now my current code for longer streams is a bit messy and unless the timings are adjusted the added delays probably will break legacy applications which also must be able to tolerate the extra offset errors at the end of the stream.

**Comment by lulol** [comment]
@fingolfin Yes, protocol 2 if inverted would work with the LIDL switches. But my idea of using the raw mode is to send longer manchester encoded data used by a keyfob switch.

I do not understand your statement:
> To elaborate: Single pulses don't make sense because a "pulse" only occurs if a signal level changes.

In my understanding, depending of the transmitter characteristics, the baseband signal generated by the encoder may or may not be transmitted correctly if certain limitations are not met, but in any case the baseband signal is encoded by level, not edges so is the actual signal level that is encoded, not the signal level changes what is transmitted specially if the modulation is OOK as most cheap switches use.

**Comment by fingolfin** [comment]
@lulol Yes, the level is relevant. Which means that the "zero duration at edge value X" transmissions you inserted (which in reality of course are a bit longer than zero, but definitely very short) have virtually no effect. All they are there for is to bridge the gap until the next proper edge value. 

Which is how I came to the conclusion that what you are transmitting looks like protocol 2 after "decoding". You did not respond to that at all.

Let me try to explain again: You suggest the "protocol
```
{ 620, {  0, 64 }, {  0,  1 }, {  1,  0 } },
```
and then the signals
```
ch0.on: 001 011 001 001 001 001 011 011 011 011 011 011 011 011 011 011 001 011 011 011 001 0
ch0.off: 001 011 001 001 001 001 011 011 011 011 011 011 011 011 011 011 011 011 011 011 011 0
```

You already grouped it helpfully in triplets, all of which are either 001 or 011. Let's see what these amount to:
001   ->   0 high, 1 low; 0 high, 1 low; 1 high, 0 low   -> removing the redundant length 0 edges leves:  1 low, 1 low, 1 high  ->  contracting these leaves  2 low, 1 high

Similarly, 011  amounts to sending  2 low, 1 high

Which is precisely protocol 2, except that the order of high and low is swapped.

Which is why I also pointed at PR #49, which is about allowing to send such signals with high and low swapped.

**Comment by fingolfin** [comment]
So, based on my understanding, I think this PR is the wrong way to go about things. But perhaps I am completly missing something, in which case I hope you can provide additional details and explanations, and I'll reconsider. Otherwise, this will be closed.

**Comment by fingolfin** [comment]
There is one importan point in this PR though: It probably was a mistake to limit the transmit function to 32 bits, because some future protocol may require more. This here, however, isn't it -- rcswitch never was meant for "raw" transfers. You don't need it for that either, as that can be done with 4-5 lines of code, no need for a library)

**Review by Totoche95 — COMMENTED** [review]
I tried this modified librairy to send 40 bits data, no success. it works if word is less than 32 bits. Is it operational ?
 send : 0x351e700dUL ,  received on RTL_433 : 35 1e 70 0d      32 bits OK
 send : 0x351e700dc0ULL , received on RTL_433 : 9e 70 0d c0      40 bits false
I tried also in decimal.
Perhaps the long long shift ? I will try a send fonction with code1, code2

Cheers
